# Lab 4: benchmark, do not believe

Team E. One aggregate query, the allocation question from Lab 1, run three ways against identical data: CSV with pandas, PostgreSQL, and DuckDB with Parquet.

The benchmark lives in the team repository at `scripts/benchmark.py`. This notebook clones it, installs PostgreSQL, and runs it.

Repository: https://github.com/zukazudo/dsai6226-data-engineering

In [1]:
# Close any DuckDB handle an earlier run left open. DuckDB allows one
# writer, and the scripts below run in their own processes.
try:
    con.close()
except NameError:
    pass

%cd /content
!rm -rf dsai6226-data-engineering
!git clone -q https://github.com/zukazudo/dsai6226-data-engineering.git
%cd /content/dsai6226-data-engineering
!pip install -q duckdb "psycopg[binary]"

/content
/content/dsai6226-data-engineering


**Observation**

The repository carries the ingester, the SQL and the benchmark. Nothing is pre-built: the warehouse is generated and deliberately not committed, so everything below runs from source.

In [2]:
# PostgreSQL, installed here because the brief names it. Nothing else in
# the project depends on it.
#
# -P pager=off matters: without it psql pipes output through less, which
# waits for a keypress that never comes and hangs the cell.
%env PAGER=cat

!apt-get -qq update > /dev/null
!apt-get -qq install -y postgresql postgresql-contrib > /dev/null
!service postgresql start

import time
time.sleep(5)

PSQL = "sudo -u postgres psql -P pager=off -q"
!{PSQL} -c "DROP DATABASE IF EXISTS adult;"
!{PSQL} -c "DROP ROLE IF EXISTS lab;"
!{PSQL} -c "CREATE ROLE lab LOGIN PASSWORD 'labpass' SUPERUSER;"
!sudo -u postgres createdb -O lab adult
!{PSQL} -d adult -c "SELECT version();"

env: PAGER=cat
W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)
 * Starting PostgreSQL 14 database server
   ...done.
                                                                 version                                                                  
------------------------------------------------------------------------------------------------------------------------------------------
 PostgreSQL 14.24 (Ubuntu 14.24-0ubuntu0.22.04.1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0, 64-bit
(1 row)



**Observation**

PostgreSQL is running inside this Colab machine, so all three engines are measured on the same hardware. That is the only way the comparison means anything: a benchmark that runs one engine on a laptop and another on a server measures the machines, not the engines.

In [3]:
# Build the warehouse from source using the Lab 3 ingester.
!python scripts/ingest.py

13:12:41  INFO    NumExpr defaulting to 2 threads.
13:12:41  INFO    run 1: adult.csv (4,104,734 bytes, sha256 250e154ed757)
13:12:44  INFO      rows read 32561
13:12:44  INFO      staged    32561  (0 already present from an earlier run)
13:12:44  INFO      rejected     0
13:12:44  INFO      loaded    32561
13:12:44  INFO      fact_person now holds 32,561 rows  (3766 ms)

load history
------------------------------------------------------------------------------
 run  source                          read  staged  rejected  loaded  status
   1  adult.csv                      32561   32561         0   32561  ok

table sizes
------------------------------------------------------------------------------
  dim_education                  16
  dim_marital_status              7
  dim_native_country             42
  dim_occupation                 16
  dim_race                        5
  dim_relationship                6
  dim_sex                         2
  dim_workclass                   9
  f

**Observation**

The benchmark runs against the cleaned warehouse and not the raw CSV. That matters: `adult.csv` still holds `?` and 99999, while the warehouse holds NULLs and flags. Comparing the raw file against the modelled data would compare different numbers and call it a benchmark.

In [4]:
# Scale 50 keeps the Colab runtime reasonable. The local run used 100.
!python scripts/benchmark.py --repeats 5 --scale 50 \
    --pg-dsn "postgresql://lab:labpass@localhost:5432/adult"

preparing inputs
  native       32,561 rows   csv     5.3 MB   parquet    0.3 MB
  scaled    1,628,050 rows   csv   272.5 MB   parquet    7.6 MB

NATIVE: 32,561 rows
  CSV + pandas       read+query    247.0 ms   query only     22.4 ms
  PostgreSQL         load          814.8 ms   query only     22.2 ms
  DuckDB + Parquet                      query          11.2 ms

  same answer as pandas:  PostgreSQL True   DuckDB True   (87 groups)

SCALED: 1,628,050 rows
  CSV + pandas       read+query   9389.7 ms   query only    551.9 ms
  PostgreSQL         load        18731.4 ms   query only    760.2 ms
  DuckDB + Parquet                      query         107.5 ms

  same answer as pandas:  PostgreSQL True   DuckDB True   (87 groups)

results written to warehouse/bench/results.json

COMPARISON TABLE

native, 32,561 rows
  engine                     size         load        query    read+query
  CSV + pandas             5.3 MB          n/a      22.4 ms        247 ms
  PostgreSQL               6.1

**Observation**

The line to read first is not a timing. It is `same answer as pandas: PostgreSQL True DuckDB True`. All three engines return the same 87 groups with the same counts and means, and the script refuses to report timings if they disagree. A number from an engine that computed a different answer is not a benchmark result.

On timings, DuckDB with Parquet is fastest at both sizes. The gap is small at the native size and large once the table grows, which is the point of running both.

## Three things that keep the numbers honest

**All three engines see identical data.** The denormalised `analytic_person` table is written once to CSV, to Parquet and to PostgreSQL. No engine gets a cleaner or smaller version of the problem.

**Load cost is separated from query cost.** pandas pays the file read on every single run. PostgreSQL and DuckDB pay it once. Reporting only per-query time flatters pandas, and reporting only total time flatters the databases, so both columns are shown.

**The scaled storage figures overstate the Parquet advantage.** The larger table is the real one replicated, so its columns are more repetitive than real data and compress better than real data would. `person_sk` and `fnlwgt` are varied per copy to blunt that, but the honest storage number is the native one.

In [5]:
import json
import pandas as pd

rows = json.load(open("warehouse/bench/results.json"))
df = pd.DataFrame(rows)
df["size_MB"] = (df.bytes / 1e6).round(1)
df["load_ms"] = df.load_ms.round(0)
df["query_ms"] = df.query_ms.round(1)
df["read_plus_query_ms"] = df.query_incl_read_ms.round(0)

for scale in df.scale.unique():
    sub = df[df.scale == scale]
    print(f"{scale}, {sub.rows.iloc[0]:,} rows")
    print(sub[["engine", "size_MB", "load_ms", "query_ms",
               "read_plus_query_ms"]].to_string(index=False))
    print()

native, 32,561 rows
          engine  size_MB  load_ms  query_ms  read_plus_query_ms
    CSV + pandas      5.3      NaN      22.4               247.0
      PostgreSQL      6.1    815.0      22.2                 NaN
DuckDB + Parquet      0.3      NaN      11.2                 NaN

scaled, 1,628,050 rows
          engine  size_MB  load_ms  query_ms  read_plus_query_ms
    CSV + pandas    272.5      NaN     551.9              9390.0
      PostgreSQL    301.2  18731.0     760.2                 NaN
DuckDB + Parquet      7.6      NaN     107.5                 NaN



**Observation**

The storage column is the one that does not depend on this machine being fast. Parquet holds the native dataset in roughly a seventeenth of the space the CSV needs, and less than a twentieth of what PostgreSQL uses once indexes and page overhead are counted.

The PostgreSQL load column is the cost nobody quotes. Before it can answer anything it has to be running, and the table has to be in it.

## Verdict

At 32,561 rows every engine answers in under 50 milliseconds, so for this dataset as it stands today speed is not what decides the question.

We would still choose **DuckDB with Parquet**, because it is the fastest of the three at both sizes, the file is seventeen times smaller than the CSV, and it needs no server running before anyone can ask a question.

PostgreSQL earns its load time and its extra disk only when several people write at once and the data has to survive a crash, and neither is true of a coursework warehouse that one analyst rebuilds from a clone in seconds.